File: middle-of-the-linked-list/middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution 
{
    public ListNode middleNode(ListNode head) 
    {
        if (head == null) // No list exists
            return head;
        
        ListNode slowP = head;
        ListNode fastP = head;
        while (fastP != null && fastP.next != null)
        {
            slowP = slowP.next;
            fastP = fastP.next.next;
        }
        return slowP;
    }
}

File: middle-of-the-linked-list/middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        '''
        Traverses a singly linked list to find the middle node
        If odd, returns the middle node
        If even, returns the rightmost middle node
        '''
        if head is None:
            return None
        fast = head
        slow = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next #Move fast 2
            slow = slow.next #Move slow 1
        return slow